zerver/actions/bots.py

```diff
@@ -69,6 +82,55 @@ def send_bot_owner_update_events(
     send_event(user_profile.realm, event, active_user_ids(user_profile.realm_id))
 
 
+def remove_bot_from_inaccessible_private_streams(
+    user_profile: UserProfile, *, acting_user: UserProfile
+) -> List[Tuple[UserProfile, Stream]]:
+    assert user_profile.bot_owner is not None
+
+    new_owner_subscribed_private_streams = get_subscribed_private_streams_for_user(
+        user_profile.bot_owner
+    )
+    new_owner_subscribed_private_stream_ids = [
+        stream.id for stream in new_owner_subscribed_private_streams
+    ]
+
+    bot_subscribed_private_streams = get_subscribed_private_streams_for_user(user_profile)
+    bot_subscribed_private_stream_ids = [stream.id for stream in bot_subscribed_private_streams]
+
+    stream_ids_to_unsubscribe = set(bot_subscribed_private_stream_ids) - set(
+        new_owner_subscribed_private_stream_ids
+    )
+    Subscription.objects.filter(
+        user_profile=user_profile, recipient__type_id__in=stream_ids_to_unsubscribe
+    ).update(active=False)
+
+    unsubscribed_streams = [
+        stream
+        for stream in bot_subscribed_private_streams
+        if stream.id in stream_ids_to_unsubscribe
+    ]
+    event_time = timezone_now()
+    event_last_message_id = get_last_message_id()
+    all_subscription_logs = [
+        RealmAuditLog(
+            realm=user_profile.realm,
+            acting_user=acting_user,
+            modified_user=user_profile,
+            modified_stream=stream,
+            event_last_message_id=event_last_message_id,
+            event_type=RealmAuditLog.SUBSCRIPTION_DEACTIVATED,
+            event_time=event_time,
+        )
+        for stream in unsubscribed_streams
+    ]
+
+    RealmAuditLog.objects.bulk_create(all_subscription_logs)
```

Comment:
Have not directly used `bulk_remove_subscriptions` here to maintain the order of events and also to make database changes first and then sending events. And also it helps us in reducing one query, since this code already adds a lot of queries.

Reply:
I think it's really risky to create these semi-duplicates of these code paths; every time we do this it ends up burning us later. 

I suspect we can make the events happen in the order we want by making better use of `transaction.on_commit`.